README.md
[![Supported Foundry Versions](https://img.shields.io/endpoint?url=https://foundryshields.com/version?url=https://github.com/paulo-roger/dynamic-shadows/releases/latest/download/module.json)](https://foundryvtt.com/packages/dynamic-shadows) [![Forge Installs](https://img.shields.io/badge/dynamic/json?label=Forge%20Installs&query=package.installs&suffix=%25&url=https%3A%2F%2Fforge-vtt.com%2Fapi%2Fbazaar%2Fpackage%2Fdynamic-shadows&colorB=0374b5)](https://forge-vtt.com/bazaar#sort=updated&package=dynamic-shadows)  
[![Latest Downloads](https://img.shields.io/github/downloads/paulo-roger/dynamic-shadows/latest/module.zip?color=blue&label=latest%20downloads)](https://github.com/paulo-roger/dynamic-shadows/releases/latest) [![Total Downloads](https://img.shields.io/github/downloads/paulo-roger/dynamic-shadows/module.zip?color=blue&label=total%20downloads)](https://github.com/paulo-roger/dynamic-shadows/releases)  
[![Discord](https://dcbadge.vercel.app/api/shield/219289132235489280?style=flat)](https://discordapp.com/users/219289132235489280) [![Ko-fi](https://img.shields.io/badge/Ko--fi-winterwulf-0374b5?logo=kofi)](https://ko-fi.com/winterwulf)

# Dynamic Shadows

[Short Description]

## Functions

## Settings

## Dependencies
- [Token Magic FX](https://foundryvtt.com/packages/tokenmagic)

## Known Issues
So far none.

## Feedback
If you have suggestions or want to report a problem, you can create an issue here: [Issues](../../issues)

## Changelog
You can read the changelog here: [CHANGELOG.md](/CHANGELOG.md)

## Special Thanks
`Peterson`, `roi007leaf` and to `SecretFire` for creating [Token Magic FX](https://foundryvtt.com/packages/tokenmagic)

## Donations
The module is totally free and will remain this way.  
I am unemployed, though. So every little help counts.

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/winterwulf)

## Attributions
--


CHANGELOG.md
# Changelog

## [0.1.0]
- Pre-release


scripts/old.js
import { MODULE, MODULE_DIR } from "./const.js"; //import the const variables
import { registerSettings, cacheSettings } from "./settings.js" //import settings

Hooks.on("preUpdateToken", async (token, updateData) => {
    let tokenX = getProperty(updateData, "x");
    let tokenY = getProperty(updateData, "y");
    let originX = token.x
    let originY = token.y

    await new Promise(resolve => setTimeout(resolve, 100));
    console.log("************novo movimento *************************", CanvasAnimation.getAnimation(`Token.${token.id}.animate`).duration)
    await new Promise(resolve =>
        setTimeout(resolve, CanvasAnimation.getAnimation(`Token.${token.id}.animate`).duration + 100))
    // .then(() => {

    if (canvas.tokens.get(token.id).light.active === true && (tokenX !== undefined || tokenY !== undefined)) {
        ui.notifications.info('esse cara tem luz')
    }

    if (tokenX !== undefined && tokenY === undefined) {
        tokenY = originY
        /* await */ createManyShadows(token, tokenX, tokenY);
    }
    else if (tokenX === undefined && tokenY !== undefined) {
        tokenX = originX
        /* await */ createManyShadows(token, tokenX, tokenY);
    } else if (tokenX !== undefined && tokenY !== undefined) {
        /* await */ createManyShadows(token, tokenX, tokenY);
    }
    // });
});

function radToDeg(rad) {
    return rad / (Math.PI / 180);
}

async function createManyShadows(token, tokenX, tokenY) {
    let canvasToken = canvas.tokens.get(token.id)
    // console.log(" ====================================== 🕵 Shadows  ======================================= ")
    let lightsInScene = canvas.scene.lights.filter(i => i.hidden === false)
    let tokensLightsInScene = canvas.tokens.objects.children.filter(i => i.light.active === true)
    for (var i = 0; i < lightsInScene.length; i++) {
        let lightX = lightsInScene[i].x
        let lightY = lightsInScene[i].y
        let catop = tokenY - lightY + (canvas.scene.grid.size / 2)
        let catad = tokenX - lightX + (canvas.scene.grid.size / 2)
        let arctan = Math.atan(catop / catad)
        let distance = Math.hypot(catop, catad) / (canvas.scene.grid.size / canvas.scene.grid.distance)
        let lightRadius = lightsInScene[i].config.dim
        arctan = radToDeg(arctan)

        // console.log(` ====================================== sombra n.${i} ======================================= `)
        // console.log(`light X: ${lightX}, light Y: ${lightY}, token X: ${tokenX}, token Y: ${tokenY}, raio da luz:${lightRadius}`)
        // console.log(`cat oposto: ${catop}, cat adjacente: ${catad}, arc tangente: ${arctan}, distancia da Luz: ${distance}`)

        if ((distance + (canvas.scene.grid.distance / 2)) <= lightRadius) {
            await castShadow(canvasToken, catad, arctan, lightsInScene[i].id)
        }
        else if (canvasToken.TMFXhasFilterId(lightsInScene[i].id)) {
            await canvasToken.TMFXdeleteFilters(lightsInScene[i].id)
        }
    }
    for (var i = 0; i < tokensLightsInScene.length; i++) {
        let lightX = tokensLightsInScene[i].light.x
        let lightY = tokensLightsInScene[i].light.y
        let catop = tokenY - lightY + (canvas.scene.grid.size / 2)
        let catad = tokenX - lightX + (canvas.scene.grid.size / 2)
        let arctan = Math.atan(catop / catad)
        let distance = Math.hypot(catop, catad) / (canvas.scene.grid.size / canvas.scene.grid.distance)
        let lightRadius = tokensLightsInScene[i].light.radius
        arctan = radToDeg(arctan)

        // console.log(` ====================================== sombra n.${i} ======================================= `)
        // console.log(`light X: ${lightX}, light Y: ${lightY}, token X: ${tokenX}, token Y: ${tokenY}, raio da luz:${lightRadius}`)
        // console.log(`cat oposto: ${catop}, cat adjacente: ${catad}, arc tangente: ${arctan}, distancia da Luz: ${distance}`)

        if ((distance + (canvas.scene.grid.distance / 2)) <= lightRadius / (canvas.scene.grid.size / canvas.scene.grid.distance)) {
            await castShadow(canvasToken, catad, arctan, tokensLightsInScene[i].id)
        }
        else if (canvasToken.TMFXhasFilterId(tokensLightsInScene[i].id)) {
            await canvasToken.TMFXdeleteFilters(tokensLightsInScene[i].id)
        }
    }
}

async function castShadow(canvasToken, catad, arctan, id) {
    let shadowAngle = arctan
    if (catad < 0) shadowAngle = arctan + 180
    let params =
        [{
            filterType: "shadow",
            filterId: id,
            rotation: shadowAngle,
            blur: 2,
            quality: 2,
            distance: 20,
            alpha: 0.6,
            padding: 10,
            shadowOnly: false,
            color: 0x000000,
            zOrder: 6000,
            animated:
            {
                blur:
                {
                    active: true,
                    loopDuration: 500,
                    animType: "syncCosOscillation",
                    val1: 2,
                    val2: 4
                },
                rotation:
                {
                    active: true,
                    loopDuration: 100,
                    animType: "syncSinOscillation",
                    val1: shadowAngle,
                    val2: shadowAngle + 4
                }
            }
        }];
    await TokenMagic.addUpdateFilters(canvasToken, params);
}

scripts/dynamic-shadows.js
import { MODULE, MODULE_DIR } from "./const.js"; //import the const variables
import { registerSettings, cacheSettings } from "./settings.js" //import settings

//Compatibility with v9
let fvttVersion

// Hook that trigger once when the game is initiated. Register and cache settings.
Hooks.once("init", () => {
    // registerWrappers();
    // registerSettings();
    // cacheSettings();
});

// hook to check the foundry version and to show the module is loaded.
Hooks.once('ready', async function () {
    fvttVersion = parseInt(game.version)
    console.log(" ====================================== 🕵 Shadows  ======================================= ")
    console.log(" ==================================== FoundryVTT Version:", fvttVersion, " ==================================== ")
});

// hook that triggers every time the light refresh, check light collision for every token in the ground level (for now lets keep it simple) and that is visible.
Hooks.on("lightingRefresh", async () => {
    let tokens = canvas.scene.tokens.contents.filter(i => i.elevation === 0 && i.hidden === false)
    for (var i = 0; i < tokens.length; i++) {
        await lightCollision(tokens[i], tokens[i].x, tokens[i].y)
    }
});

// convert radians to degrees
function radToDeg(rad) {
    return rad / (Math.PI / 180);
}

// calc the sides of the triangle
function cathetus(tokenXY, lightXY) {
    return tokenXY - lightXY + (canvas.scene.grid.size / 2)
}

// hypothenuse + adjust using canvas grid properties to use the center point of the square. Using the grid properties makes the code adjustable to any grid size or unit
function calcDistance (cathetusA, cathetusB){
   return (Math.hypot(cathetusA, cathetusB) / (canvas.scene.grid.size / canvas.scene.grid.distance)) + (canvas.scene.grid.distance / 2)
}

// checks light collision with the token
async function lightCollision(token, tokenX, tokenY) {
    let canvasToken = canvas.tokens.get(token.id)
    // console.log(" ====================================== 🕵 DEBUG: Shadows  ======================================= ")
    let lightsInScene = canvas.scene.lights.filter(i => i.hidden === false) // array of lights in the scene that are not faded.
    let tokensLightsInScene = canvas.tokens.objects.children.filter(i => i.light.active === true) // array of tokens that are emitting light.
    for (var i = 0; i < lightsInScene.length; i++) {
        // console.log(" ====================================== 🕵 DEBUG: lightingRefresh - LIGHTS 💡  ======================================= ")
        let lightX = lightsInScene[i].x
        let lightY = lightsInScene[i].y
        let OpoCathetus = cathetus(tokenY, lightY)
        let adjCathetus = cathetus(tokenX, lightX)
        let distance = calcDistance(OpoCathetus, adjCathetus) //distance of the center of the token square to the center of the light square
        let lightRadius = lightsInScene[i].config.dim

        // console.log(` ====================================== sombra n.${i} ======================================= `)
        // console.log(`light X: ${lightX}, light Y: ${lightY}, token X: ${tokenX}, token Y: ${tokenY}, raio da luz:${lightRadius}`)
        // console.log(`cat oposto: ${OpoCathetus}, cat adjacente: ${adjCathetus}, distancia da Luz: ${distance}`)

        if ((distance) <= lightRadius) {
            await castShadow(canvasToken, adjCathetus, OpoCathetus, lightsInScene[i].id)
            // console.log(" ====================================== 🕵 DEBUG: REFRESHING SHADOW  ======================================= ")
        }
        else if (canvasToken.TMFXhasFilterId(lightsInScene[i].id)) {
            // console.log(" ====================================== 🕵 DEBUG: REMOVING SHADOW  ======================================= ")
            await canvasToken.TMFXdeleteFilters(lightsInScene[i].id)
        }
    }
    for (var i = 0; i < tokensLightsInScene.length; i++) {
        // console.log(" ====================================== 🕵 DEBUG: lightingRefresh - TOKENS ⛹ ======================================= ")
        let lightX = tokensLightsInScene[i].light.x
        let lightY = tokensLightsInScene[i].light.y
        let OpoCathetus = cathetus(tokenY, lightY)
        let adjCathetus = cathetus(tokenX, lightX)
        let distance = calcDistance(OpoCathetus, adjCathetus)
        let lightRadius = tokensLightsInScene[i].light.radius

        // console.log(` ====================================== sombra n.${i} ======================================= `)
        // console.log(`light X: ${lightX}, light Y: ${lightY}, token X: ${tokenX}, token Y: ${tokenY}, raio da luz:${lightRadius}`)
        // console.log(`cat oposto: ${OpoCathetus}, cat adjacente: ${adjCathetus}, distancia da Luz: ${distance}`)

        //this if is for the shadow of the torch holder, creates a shadow below her/him.
        if (token.id == tokensLightsInScene[i].id) {
            if (!canvasToken.TMFXhasFilterId(token.id)) {
                await castSelfShadow(canvasToken, tokensLightsInScene[i].id)
            }
        }
        else {
            if ((distance) <= lightRadius) {
                // console.log(" ====================================== 🕵 DEBUG: REFRESHING SHADOW  ======================================= ")
                await castShadow(canvasToken, adjCathetus, OpoCathetus, tokensLightsInScene[i].id)
            }
            else if (canvasToken.TMFXhasFilterId(tokensLightsInScene[i].id)) {
                await canvasToken.TMFXdeleteFilters(tokensLightsInScene[i].id)
                // console.log(" ====================================== 🕵 DEBUG: REMOVING SHADOW  ======================================= ")
            }
        }
    }
}

// creates the shadow with the parameters params below.
async function castShadow(canvasToken, adjCathetus, OpoCathetus, id) {
    let shadowAngle = Math.round(radToDeg(Math.atan(OpoCathetus / adjCathetus))) // angle of the shadow around the token.
    if (adjCathetus < 0) shadowAngle += 180 // adjust the angle in case of the adjacent cathetus is  negative
    console.log(` ====================================== SHADOW ANGLE: ${shadowAngle} ======================================= `)
    let params =
        [{
            filterType: "shadow",
            filterId: id,
            rotation: shadowAngle,
            blur: 2,
            quality: 4,
            distance: (canvas.scene.grid.size / 2),
            alpha: 0.6,
            padding: 10,
            shadowOnly: false,
            color: 0x000000,
            zOrder: 6000,
            animated:
            {
                blur:
                {
                    active: true,
                    loopDuration: 500,
                    animType: "syncCosOscillation",
                    val1: 2,
                    val2: 3
                },
                rotation: //makes the shadow flickers
                {
                    active: true,
                    loopDuration: 300,
                    animType: "syncSinOscillation",
                    val1: shadowAngle,
                    val2: shadowAngle + 4
                }
            }
        }];
    if (canvasToken.document.getFlag('tokenmagic', 'filters') !== undefined) { // this part is to optmize the code, if the shadow already exists and the angle didn't change, do nothing. else apply the shadow filter.
        if (canvasToken.document.getFlag('tokenmagic', 'filters').find(i => i.tmFilters.tmParams.rotation === shadowAngle) === undefined)
            await TokenMagic.addUpdateFilters(canvasToken, params);
    } else await TokenMagic.addUpdateFilters(canvasToken, params);
}

async function castSelfShadow(canvasToken, id) {
    let shadowAngle = 90
    let params =
        [{
            filterType: "shadow",
            filterId: id,
            rotation: shadowAngle,
            blur: 2,
            quality: 4,
            distance: (canvas.scene.grid.size / 3),
            alpha: 0.5,
            padding: 10,
            shadowOnly: false,
            color: 0x000000,
            zOrder: 6000,
            animated:
            {
                blur:
                {
                    active: true,
                    loopDuration: 500,
                    animType: "syncCosOscillation",
                    val1: 2,
                    val2: 3
                },
                rotation:
                {
                    active: true,
                    loopDuration: 200,
                    animType: "syncSinOscillation",
                    val1: shadowAngle,
                    val2: shadowAngle + 4
                }
            }
        }];
    await TokenMagic.addUpdateFilters(canvasToken, params);
}